upf/upf_pfcp_api.c

```diff
@@ -2892,9 +2892,32 @@ handle_session_report_request (pfcp_msg_t * msg, pfcp_decoded_msg_t * dmsg)
   return -1;
 }
 
+/**
+ * @brief Handle a PFCP Session Report Response
+ *
+ * @note dmsg has to contain valid SEID
+*/
 static int
 handle_session_report_response (pfcp_msg_t * msg, pfcp_decoded_msg_t * dmsg)
 {
+  if (dmsg->session_report_response.response.cause ==
+      PFCP_CAUSE_SESSION_CONTEXT_NOT_FOUND)
+    {
+      // control plane does not know about the session. drop it
+      upf_main_t *gtm = &upf_main;
+      upf_session_t *sess;
+
+      if (!(sess = pool_elt_at_index (gtm->sessions, msg->session_index)))
```

Comment:
`pool_elt_at_index()` contains `ASSERT`, ensuring the item is not free, so in order to avoid a crash you need to check `pool_is_free_index()` beforehand. `pool_elt_at_index` never returns a null pointer.

There's another concern, namely that there may be an edge case with the following sequence of events:
1. UPG sends PFCP Session Report Response to the SMC, but it doesn't arrive immediately
2. The session is deleted by the SMC
3. A new session is created by the SMC and takes up the same index in the pool as a prev session
4. The Session Report Response gets to the SMC and is replied to with Session context not found, as the session has already been deleted
5. A late Session Report response is sent from SMC towards UPG with Session context not found error
6. UPG handling this report will cause the session created in p.3 to be deleted, even though it's fully ok

As a safeguard, we could probably add `u64 seid` field to the `pfcp_msg_t` and compare it to `sess->cp_seid` after we get the session by its index